### ISSUE TITLE
Question / Comment: What is `page.xref`

### ISSUE THREAD

**Issue by sant527** [pymupdf/PyMuPDF#633]
I was digging into the `setToC` function in `utils.py`.

Is there any shortcut way to get `page.xref` 

currently it uses:

 `page=doc[pno] and then get page.xref`

**Comment by JorjMcKie** [comment]
Yes, the base function if this method is ``doc._getPageObjNumber(pno)``. It delivers the integer tuple ``(xref, gen)`` **_without reading the page_** (gen = generation number, mostly zero). So the xref can be gotten via ``xref = doc._getPageObjNumber(pno)[0]``.
However, if you need the page height, you  must load the page nonetheless ...

**Comment by sant527** [comment]
Thats very helpful. With the help of this i was able to speed up the `setToC()` for my requirement.

So the following solved my problem for `setToc` slow when the `toc` consists of 50000+ entries. 

For my requirement its only bookmark pointing to the top of the page

So there is no need to calculate the height of the page. So not calling the the `page[pno]` significantly reduces the time for `setToc`

This question i have asked in https://github.com/pymupdf/PyMuPDF/issues/623

If i use 
```
dest_dict = {"to": Point(0, 0), "kind": LINK_GOTO}
```

instead of 

```
page = doc[pno]  # load the page
ictm = ~page.transformationMatrix  # get inverse transformation matrix
top = Point(72, 36) * ictm  # default top location
dest_dict = {"to": top, "kind": LINK_GOTO}  # fall back target
```

and 
```
d["dest"] = getDestStr(doc._getPageObjNumber(pno)[0], dest_dict)
```

instead of 

```
d["dest"] = getDestStr(page.xref, dest_dict)
```

**Comment by JorjMcKie** [comment]
Hey, congratulations!
Your input gives me reason to revisit the TOC stuff for performance again.
What I understand is that this setToC version is faster. What I do not yet understand: why would the PDF **_open_** faster??!!
I'll check that out.

**Comment by JorjMcKie** [comment]
Thanks to your input, I have created faster utility methods for extracting xref (existed before) and rectangle of a page:
Both now work **_without loading the page_**.
Also changed ``setToC`` method to use them. At least for medium-sized TOC replacements (like the 800 entries in the Adobe manual) I observed a more than **_ten times higher speed:_** 0.06 seconds instead of 0.8 before.

For those absolute giant quantities you are dealing with, this will probably not be the same (because other influences come into play like storage acquisition), but a factor of two times better should be expected.
